src/hivpy/sexual_behaviour.py

```diff
@@ -277,3 +284,99 @@ def update_rred_balance(self, population):
+        ltp_rands = rng.random(size=size)
+        rate_modifier = 1
+        if (35 <= age < 45):
+            rate_modifier = 0.5
+        elif (45 <= age < 55):
+            rate_modifier = 1/3
+        else:
+            rate_modifier = 0.2
+        new_relationship = ltp_rands < (self.new_ltp_rate
+                                        * rate_modifier
+                                        * self.ltp_balance_factor[sex])
+        return new_relationship
+
+    def new_ltp_longevity(self, age, sex, size):
+        longevity = np.zeros(size)  # each new relationship needs a longevity
+        longevity_rands = rng.random(size)
+        if (age < 45):
+            longevity = ((longevity_rands < 0.3) * 1 +
+                         ((0.3 <= longevity_rands) & (longevity_rands < 0.6)) * 2 +
+                         (0.6 <= longevity_rands) * 3)
+        elif (age < 55):
+            longevity = ((longevity_rands < 0.3) * 1 +
+                         ((0.3 <= longevity_rands) & (longevity_rands < 0.8)) * 2 +
+                         (0.8 <= longevity_rands) * 3)
+        else:
+            longevity = ((longevity_rands < 0.3) * 1 +
+                         (0.3 <= longevity_rands) * 2)
+        return longevity
+
+    def continue_ltp(self, age, longevity, size):
+        """Function to decide which long term partners cease condomless sex based on
+        relationship longevity, age, and sex."""
+        # TODO: Add balancing factors for age and sex demographics.
+        end_probability = self.ltp_end_rate_by_longevity[int(longevity)] / self.ltp_rate_change
+        r = rng.random(size=size)
+        return (r > end_probability)
+
+    def update_long_term_partners(self, population):
+        start_partnerless = ((~population.data[col.LONG_TERM_PARTNER]) &
+                             (population.data[col.AGE] >= 15) &
+                             (population.data[col.AGE] < 65))
+        partnerless_idx = population.data.index[start_partnerless]
+
+        partnered_idx = population.data.index[(population.data[col.LONG_TERM_PARTNER]) &
+                                              (population.data[col.AGE] >= 15) &
+                                              (population.data[col.AGE] < 65)]
+
+        (self.ltp_balance_factor[SexType.Male],
+         self.ltp_balance_factor[SexType.Female]) = self.balance_ltp_factor(population)
+
+        # new relationships
+        new_relationships = population.transform_group(
+            [col.SEX, col.AGE],  self.start_new_ltp, sub_pop=partnerless_idx)
+        population.data.loc[partnerless_idx, col.LONG_TERM_PARTNER] = new_relationships
+        # not sure about indexing on this next line
+        # We only want to update LTP_LONGEVITY where LONG_TERM_PARTNER is true within
+        # the sub-population defined by partnerless_idx
+        new_ltp_subpop = population.data.index[start_partnerless &
+                                               population.data[col.LONG_TERM_PARTNER]]
```

Comment:
The indexing looks okay to me: they were partnerless at the start and now have a partner